src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/ConsiderPassingBaseTypesAsParameters.cs

```diff
@@ -0,0 +1,227 @@
+                        foreach (var (parameter, usedTypesPool) in typesUsedPerParameter)
+                        {
+                            var usedTypes = usedTypesPool.ToImmutableArray();
+                            // If type is unused, we ignore it. There is another rule to detect
+                            // unused parameters.
+
+                            if (usedTypes.Length == 1)
+                            {
+                                ReportIfTypesAreDifferent(parameter, usedTypes[0]);
+                            }
+                            else if (usedTypes.Length > 1)
+                            {
+                                ReportIfTypesAreDifferent(parameter, FindMostGenericType(parameter.Type, usedTypes));
+                            }
+                        }
+
+                        // Local functions
+                        void ReportIfTypesAreDifferent(IParameterSymbol parameter, ITypeSymbol suggestedBaseTypeSymbol)
+                        {
+                            if (!SymbolEqualityComparer.Default.Equals(suggestedBaseTypeSymbol, parameter.Type))
+                            {
+                                // All calls on this parameter are using the same base type which is different
+                                // from the one declared, so we can simply report.
+                                context.ReportDiagnostic(parameter.CreateDiagnostic(Rule, parameter.Name,
+                                    parameter.Type.Name, suggestedBaseTypeSymbol.Name));
+                            }
+                        }
+                    });
+                });
+            });
+        }
+
+        private static ITypeSymbol FindDeclaringType(ISymbol symbol)
+        {
+            Debug.Assert(
+                symbol.Kind != SymbolKind.Method ||
+                symbol.Kind != SymbolKind.Property ||
+                symbol.Kind != SymbolKind.Event);
+
+            if (symbol.IsOverride)
+            {
+                return FindDeclaringType(symbol.GetOverriddenMember());
+            }
+
+            var interfaceMembers = ExplicitOrImplicitInterfaceImplementations(symbol);
+            if (interfaceMembers.Length == 1)
+            {
+                return interfaceMembers[0].ContainingType;
+            }
+            else
+            {
+                // When reaching this point, either we found no interface member, which means the
+                // member is declared directly on the type OR more than one interface was found
+                // as declaring this type so let's assume it is defined on the containing type.
+                return symbol.ContainingType;
+            }
+        }
+
+        // Borrowed from Microsoft.CodeAnalysis.Workspaces/ISymbolExtensions.cs with a small tweak
+        // which consist in passing the member name.
+        private static ImmutableArray<ISymbol> ExplicitOrImplicitInterfaceImplementations(ISymbol symbol)
```

Comment:
I am surprised I couldn't find such helper in the codebase, have I missed something?

Reply:
Can this be moved to ISymbolExtensions.cs?